### PR TITLE
Implement project generation

### PR DESCRIPTION
**Pull request by janosbrodbeck**
~~**PR is based on #49, so the other PR has to be merged first.** Only commits from 04.07 are from this PR.~~
* Add `language` and `file_extension` field to `GenerationTemplate`
Language for filtering and easier viewing the templates. Extension required for creating the source file, as we cannot implement all file endings for potential template languages
* `Robot` got a relationship to welding point
* Template rendering does not strip trailing newlines anymore
* Endpoint for generating the project:
Files and zip file are created completely in-memory

Testing is hard, as a test is missing and frontend is not completely working for this branch. Test is on the to-do list, but regarding other tasks for the project a bit lower in priority right now. I prepared my working environment like this:
1) Create database from main
2) Add project, upload project, create robots, assign robots with the current main frontend
3) Switch to branch of this PR and comment out `op.execute("TRUNCATE robottype, robot, weldingpoint, generationtemplate;")` in the new migration!
4) In DataGrip/PGAdmin: Add template id to robot type, add langauge and extension to template
5) Run endpoint in openAPI

Used your niryo template for testing:
```
from niryo_one_python_api.niryo_one_api import NiryoOne
import rospy
import time

def grab_release_reset(niryo):
    niryo.grab_with_tool(niryo.get_current_tool_id())  # grab
    niryo.release_tool(niryo.get_current_tool_id())  # release
    niryo.move_pose(0.1, 0.1, 0.4, 0, 0, 0)  # reset


if __name__ = '__main__':
    rospy.init_node('niryo_one_example_python_api')
    n = NiryoOne()

    {% for p in welding_points -%}
    n.move_pos(x={{p.x}}, y={{p.y}}, z={{p.z}}, roll={{p.roll}}, pitch={{p.pitch}}, yaw={{p.yaw}})
    grab_release_reset(niryo)
    {%- if loop.index != welding_points|length %}
    {% endif %}{% endfor %}
```

### CODE REVIEW

File: src/app/api/api_v1/endpoints/project.py [hunk 1]
```diff
@@ -38,6 +42,40 @@ async def read_project(
...
 
 
+@router.get("/{id}/generate", response_description="Zip file containing the generated code")
+async def generate_project(
+        *,
+        db: AsyncSession = Depends(deps.get_async_db),
+        _id: int
+) -> Any:
+    """"
+    Retrieve generated code of the project
```
Comment (dolevi-tub) [inline src/app/api/api_v1/endpoints/project.py:52]:
```suggestion
    Generate and retrieve code of the project
```
This sounds too much like it assumes pre-generated code.

**Review by dolevi-tub — CHANGES_REQUESTED** [review]
Looks (and feels) better than my previous solution :smiley:

**Comment by dolevi-tub** [comment]
The both `PUT /weldingpoint/` endpoints currently don't ensure unique welding_order indexes.
With the current frontend this shouldn't be an issue but I feel we should either
* ensure it
* mention the assumption in a code comment

File: src/app/codegen/code_generator.py [hunk 2]
```diff
@@ -1,10 +1,52 @@
...
+from app.models.robot import Robot
+from app.models.welding_point import WeldingPoint
 
 
 class CodeGenerator:
     @staticmethod
-    def generate(template: GenerationTemplate, project: Project) -> str:
-        return Template(template.content).render({"welding_points": project.welding_points})
+    def generate(template: GenerationTemplate, welding_points: List[WeldingPoint]) -> str:
+        return Template(template.content, keep_trailing_newline=True).render({"welding_points": welding_points})
```
[the remark this answers is not in the document]
Comment (janosbrodbeck) [reply src/app/codegen/code_generator.py:18]: It was my intention to put it into this function, but we need a Python 3.11 feature for that, and Python 3.11 releases in october this year :( I cannot hint `Self` as return type, if its the class itself. This feature is coming with 3.11. 
E.g.
`def generate_code_by_robot() -> List[Self]` https://peps.python.org/pep-0673/

Maybe I have an idea for a workaround. Will try out later!

[the remark this answers is not in the document]
Comment (janosbrodbeck) [reply src/app/codegen/code_generator.py]: Addressed

**Pull request merged**